compiler/rustc_abi/src/layout.rs

```diff
@@ -621,7 +621,7 @@ where
     let discr_type = repr.discr_type();
     let bits = Integer::from_attr(dl, discr_type).size().bits();
     for (i, mut val) in discriminants {
-        if variants[i].iter().any(|f| f.abi.is_uninhabited()) {
+        if !repr.c() && variants[i].iter().any(|f| f.abi.is_uninhabited()) {
```

Comment:
Don't we also need to ensure that `needs_dist` above does not skip this variant? Not quite sure how to hit that code path, but at least needs a test

Reply:
I think `needs_disc` is only relevant for the niche filling layout? `#[repr(C)]` never uses niche-filling, so it shouldn’t apply here.

Reply:
We’ll hit `if dont_niche_optimize_enum` at the very beginning: it’s populated from `def.repr().inhibit_enum_layout_opt()`, which is `self.c() || self.int.is_some()`.

I’m not sure what the test would look like here: we already test that uninhabited variants get discriminants, so if we somehow reach `is_disc`, `absent(&variants[index])` would return `true` and we’ll catch this case.

Reply:
thanks! good analysis.